---
title: "Problem Set 3. Due Thurs March 2 5pm"
author: "Christy Smith"
date: "March 1, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```

# Submission

1. Fork this repository to your own account
1. Make changes to the `README.Rmd` file (including the author field above).
1. Knit the file to HTML format and add it to your git repository (`git add`)
1. Submit a pull request with your Rmd and knitted HTML files.

# Overview 

You will examine a [time-course of mRNA expression][1] of barley powdery mildew fungus in immunocompromised plants (*Arabidopsis*). The fungus Blumeria graminis f. sp. hordei (called *Bgh* from now on)

Counts were generated using the [`maSigPro` software][3] and deposited in [NCBI GEO][2].

[1]: http://www.ncbi.nlm.nih.gov/pubmed/23175786
[2]: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43163  
[3]: http://bioinformatics.oxfordjournals.org/content/30/18/2598.full

Some important identifiers for the data set:

| Name | Description |
| ---- | ----------- |
| `A6` | *Bgh* isolate expressing other AVRA effectors |
| `K1` | *Bgh* isolate expressing the cognate AVRA1 effector for MLA1|
| `pps` | *Arabidopsis* plants |
| `B12` | *Arabidopsis* plants expressing MLA1-HA|

We will download tables from GEO containing counts of mRNA abundance from both the fungal sample.

# Raw data

First we need load the data.

```{r load_raw_data, message = FALSE}
library(tidyverse)

# bgh data
bgh_url <- 'http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43163&format=file&file=GSE43163_CompleteCountTable_Bgh.txt.gz'

# 2 info and blank lines at top, skip them
raw_data <- read_tsv(bgh_url, skip = 2)

# the header for the first column is "NA", set it manually
names(raw_data)[1] <- 'gene.name'

raw_data <- as_data_frame(raw_data)
```

# Problems


## Problem 1

Now that the raw data are loaded, your first task is to tidy the data with `tidyr`. The `cleaned_data` should look something like this:

```
# A tibble: 310,896 Ã— 6
             gene.name plant fungus time.point   rep value
*                <chr> <chr>  <chr>      <chr> <chr> <int>
1             bgh04079   B12     A6       6hpi     1    13
2             bgh01634   B12     A6       6hpi     1    31
3  bghG000012000001001   B12     A6       6hpi     1   121
4  bghG000012000002001   B12     A6       6hpi     1     3
5             bgh00757   B12     A6       6hpi     1   253
6             bgh01273   B12     A6       6hpi     1    45
7             bgh01274   B12     A6       6hpi     1    32
8             bgh01277   B12     A6       6hpi     1     9
9             bgh06140   B12     A6       6hpi     1    47
10            bgh05774   B12     A6       6hpi     1    42
# ... with 310,886 more rows
```

The key steps are to `gather()` the data into key-value pairs and then `separate()` the information from the key column into new columns.

```{r problem_1}
raw_data %>% gather(key, value, -gene.name) -> raw_data_gathered

raw_data_gathered %>% separate(key, into=c('plant', 'fungus', 'time.point', 'rep'), sep='_') -> raw_data_separated

#Now we have tidy data:
raw_data_separated

```

## Problem 2

You need to translate what you learned on the command line into the R world.

Which plant has the highest expression of any gene in the 6hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
```

```{r}
library(stringr)

raw_data_separated %>% mutate(time.point = str_replace(time.point, 'hpi', ''), time.point = str_c('hpi.', time.point)) -> hpi_dot

hpi_dot %>% spread(time.point, value) -> cleaned_data

cleaned_data %>% arrange(desc(hpi.6)) %>% slice(1) %>% select(plant)

```

Which plant / fungus pair has the highest expression in the 18hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
```

```{r}
cleaned_data %>% arrange(desc(hpi.18)) %>% slice(1) %>% select(plant,fungus)
```

## Problem 3

Identify the top 3 most consistently differentially expressed genes between the earliest and latest time points for each combination of `plant` and `fungus` strains.

- "Differential expression"" is the difference between `value` (i.e., gene expression level) between time points.

- "Consistency" is the smallest variance in `value` between replicates.

### Strategy

1. Create a new table from the cleaned data by moving each `hpi` value to a new column name, with counts for each in the column (hint: use a `tidyr` verb). It is helpful to reformat the `hpi` values by converting from e.g. `6hpi` to `hpi.6`. You can use `mutate` to do this, i.e.:

```{r reformat_time, eval=FALSE}
cleaned_data %>%
  mutate(time.value = str_replace(time.point, 'hpi', ''),
         time.hpi = str_c('hpi.', time.value))
```

2. Create a new column containing the expression difference between the relevant time points.

3. Calculate summary statistics (mean and variance) of the expression differences by grouping (hint) the `gene.name`, `plant`, and `virus` columns.

4. Sort by these statistics and use the `dplyr` verb `slice` to pull the ones you want (i.e., the top 3). Note you will have to remove `gene.name` from the grouping so that sorting works.

```{r problem_3}
#First we find the difference between the earliest and latest time points and put the difference in a new column called expr.diff: 
cleaned_data %>% mutate(expr.diff = hpi.24 - hpi.6) -> expr_data

#Next we group the data by gene name, plant, and fungus, which results in 3 different values of expr.diff for each group. Then we find the mean and variance for these 3 values of expr.diff for each group. Next we will re-group the data by just plant and fungus to display these means and variances. We will arrange the data to find the highest mean expr.diff and then use the lowest variance of expr.diff as secondary sorting criteria (ie, if there was a tie between two means). Finally, we will report the top 3 expr.diff values that meet this criteria for each plant/fungus group. 
expr_data %>% group_by(gene.name, plant, fungus) %>% summarise(mean.expr.diff = mean(expr.diff), var.expr.diff = var(expr.diff)) %>% group_by(plant, fungus) %>% arrange(desc(abs(mean.expr.diff)), var.expr.diff) %>% slice(1:3) -> top3_diff_expr_genes

#The final data frame is:
top3_diff_expr_genes

```

## Problem 4

Now examine the above final data frame above and write a few sentences putting inline code in least 3 places. For example, There are `r nrow(mtcars)` rows of data in `mtcars`.

We created a data frame with `r nrow(top3_diff_expr_genes)` rows and `r ncol(top3_diff_expr_genes)` columns. The column names of this dataframe are:
`r colnames(top3_diff_expr_genes)`.The highest variance in this data frame is `r max(top3_diff_expr_genes$var.expr.diff)` and the lowest variance is `r min(top3_diff_expr_genes$var.expr.diff)`. 

## Problem 5

Plot the expression (`value`) by time (`hpi`) for the above genes. Format the plot as follows:

1. Plot each point.
1. Connect the points with a smooth line
1. Create a facet for each `plant` / `bgh` pair 
1. Assign a different color to each gene
1. Assign a different shape to each replicate  
1. Add a descriptive title to the plot
1. Use the "black & white" theme

Extra credit: add error bars to the plot (use `geom_errorbar`).

```{r problem_5}

ggplot(subset(hpi_dot, gene.name %in% top3_diff_expr_genes$gene.name)) + geom_point(aes(x=time.point, y=value, color=factor(gene.name), shape=factor(rep))) + geom_smooth(method='lm', se=FALSE, aes(x=time.point, y=value, group=interaction(gene.name, rep))) + facet_wrap(plant ~fungus) + ggtitle('mRNA Expression Time-Course for Bgh') + theme_bw()


```
